domain/domain.go

```diff
@@ -1221,6 +1245,128 @@ func (do *Domain) NotifyUpdatePrivilege(ctx sessionctx.Context) {
 	}
 }
 
+// ServerID gets serverID.
+func (do *Domain) ServerID() uint64 {
+	return atomic.LoadUint64(&do.serverID)
+}
+
+func (do *Domain) getServerIDSession() *concurrency.Session {
+	s, _ := do.serverIDSession.Load().(*concurrency.Session)
+	return s
+}
+
+const (
+	serverIDEtcdPath             = "/tidb/server_id"
+	acquireServerIDRetryCnt      = 100
+	acquireServerIDRetryInterval = 300 * time.Millisecond
+	acquireServerIDTimeout       = 3 * time.Second
+)
+
+func (do *Domain) acquireServerID(ctx context.Context) error {
+	cfg := config.GetGlobalConfig()
+
+	atomic.StoreUint64(&do.serverID, 0)
+
+	logPrefix := fmt.Sprintf("[acquireServerID] ")
+	session, err := owner.NewSession(ctx, logPrefix, do.etcdClient, owner.NewSessionDefaultRetryCnt, cfg.ServerIDTTL)
+	if err != nil {
+		return err
+	}
+	do.serverIDSession.Store(session)
+
+	for i := 0; i < acquireServerIDRetryCnt; i++ {
+		randServerID := rand.Int63n(int64(util.MaxServerID)) + 1 // get a random serverID: [1, MaxServerID]
+		key := fmt.Sprintf("%s/%v", serverIDEtcdPath, randServerID)
+		cmp := clientv3.Compare(clientv3.CreateRevision(key), "=", 0)
+		value := "0"
+
+		childCtx, cancel := context.WithTimeout(ctx, acquireServerIDTimeout)
+		txn := do.etcdClient.Txn(childCtx)
+		t := txn.If(cmp)
+		resp, err := t.Then(clientv3.OpPut(key, value, clientv3.WithLease(do.getServerIDSession().Lease()))).Commit()
+		cancel()
+		if err != nil {
+			return err
+		}
+		if !resp.Succeeded {
+			logutil.BgLogger().Info("random serverID exists, try again", zap.Int64("randServerID", randServerID))
+			time.Sleep(acquireServerIDRetryInterval)
+			continue
+		}
+
+		atomic.StoreUint64(&do.serverID, uint64(randServerID))
+		logutil.BgLogger().Info("acquireServerID", zap.Uint64("serverID", do.ServerID()),
+			zap.String("lease id", strconv.FormatInt(int64(do.getServerIDSession().Lease()), 16)))
+		return nil
+	}
+	logutil.BgLogger().Warn("no serverID available", zap.Int32("retry count", acquireServerIDRetryCnt))
+	return nil
+}
+
+func (do *Domain) refreshServerIDTTL(ctx context.Context) error {
+	key := fmt.Sprintf("%s/%v", serverIDEtcdPath, do.ServerID())
+	value := "0"
+	err := ddlutil.PutKVToEtcd(ctx, do.etcdClient, acquireServerIDRetryCnt, key, value, clientv3.WithLease(do.getServerIDSession().Lease()))
+	if err != nil {
+		logutil.BgLogger().Error("refreshServerIDTTL fail", zap.Uint64("serverID", do.ServerID()), zap.Error(err))
+	} else {
+		logutil.BgLogger().Info("refreshServerIDTTL succeed", zap.Uint64("serverID", do.ServerID()))
+	}
+	return err
+}
+
+func (do *Domain) serverIDKeeper() {
+	cfg := config.GetGlobalConfig()
+	serverIDTimeToCheckPDConnection := time.Duration(cfg.ServerIDTTL) * time.Second / 10
+	lostConnectionTimeout := serverIDTimeToCheckPDConnection * 5
```

Comment:
We must find a way to ensure that `lostConnectionTimeout` is actually less than `cfg.ServerIDTTL`.
Because the server should `KillAllConnections` before the other tidb acquires the same connID.

Reply:
And because tidb will `KillAllConnections` after it thinks `lostConnectionTimeout` is timeout, for not barbarically killing the on-going long-run SQL, so I think the default `lostConnectionTimeout` and `cfg.ServerIDTTL` should be large enough. @breeswish suggests that `cfg.ServerIDTTL` should default to 12 hours.